process/block/preprocess/miniBlocksCompaction.go

```diff
@@ -0,0 +1,289 @@
+	shardCoordinator sharding.Coordinator
+
+	mapHashesAndTxs         map[string]data.TransactionHandler
+	mapSenderNonce          map[string]uint64
+	mapUnallocatedTxsHashes map[string]struct{}
+
+	mutMiniBlocksCompaction sync.RWMutex
+}
+
+// NewMiniBlocksCompaction creates a new mini blocks compaction object
+func NewMiniBlocksCompaction(
+	economicsFee process.FeeHandler,
+	shardCoordinator sharding.Coordinator,
+) (*miniBlocksCompaction, error) {
+
+	if economicsFee == nil || economicsFee.IsInterfaceNil() {
+		return nil, process.ErrNilEconomicsFeeHandler
+	}
+	if shardCoordinator == nil || shardCoordinator.IsInterfaceNil() {
+		return nil, process.ErrNilShardCoordinator
+	}
+
+	mbc := miniBlocksCompaction{
+		economicsFee:     economicsFee,
+		shardCoordinator: shardCoordinator,
+	}
+
+	mbc.mapHashesAndTxs = make(map[string]data.TransactionHandler)
+	mbc.mapSenderNonce = make(map[string]uint64)
+	mbc.mapUnallocatedTxsHashes = make(map[string]struct{})
+
+	return &mbc, nil
+}
+
+// Compact method tries to compact the given mini blocks to have only one mini block per sender/received pair
+func (mbc *miniBlocksCompaction) Compact(
+	miniBlocks block.MiniBlockSlice,
+	mapHashesAndTxs map[string]data.TransactionHandler,
+) block.MiniBlockSlice {
+
+	mbc.mutMiniBlocksCompaction.Lock()
+	defer mbc.mutMiniBlocksCompaction.Unlock()
+
+	if len(miniBlocks) <= 1 {
+		return miniBlocks
+	}
+
+	mbc.mapHashesAndTxs = mapHashesAndTxs
+
+	compactedMiniBlocks := make(block.MiniBlockSlice, 0)
+	compactedMiniBlocks = append(compactedMiniBlocks, miniBlocks[0])
+
+	for index, miniBlock := range miniBlocks {
+		if index == 0 {
+			continue
+		}
+
+		compactedMiniBlocks = mbc.merge(compactedMiniBlocks, miniBlock)
+	}
+
+	if len(miniBlocks) > len(compactedMiniBlocks) {
+		log.Info(fmt.Sprintf("compacted from %d miniblocks to %d miniblocks\n",
+			len(miniBlocks), len(compactedMiniBlocks)))
+	}
+
+	return compactedMiniBlocks
+}
+
+func (mbc *miniBlocksCompaction) merge(
+	mergedMiniBlocks block.MiniBlockSlice,
+	miniBlock *block.MiniBlock,
+) block.MiniBlockSlice {
+
+	for _, mergedMiniBlock := range mergedMiniBlocks {
+		sameType := miniBlock.Type == mergedMiniBlock.Type
+		sameSenderShard := miniBlock.SenderShardID == mergedMiniBlock.SenderShardID
+		sameReceiverShard := miniBlock.ReceiverShardID == mergedMiniBlock.ReceiverShardID
+
+		canMerge := sameSenderShard && sameReceiverShard && sameType
+		if canMerge {
+			haveEnoughGasToMerge := mbc.haveEnoughGasToMerge(mergedMiniBlock, miniBlock)
+			if haveEnoughGasToMerge {
+				mergedMiniBlock.TxHashes = append(mergedMiniBlock.TxHashes, miniBlock.TxHashes...)
+				return mergedMiniBlocks
+			}
+		}
+	}
+
+	mergedMiniBlocks = append(mergedMiniBlocks, miniBlock)
+
+	return mergedMiniBlocks
+}
+
+func (mbc *miniBlocksCompaction) haveEnoughGasToMerge(
+	destMiniBlock *block.MiniBlock,
+	srcMiniBlock *block.MiniBlock,
+) bool {
+
+	gasUsedInDestMiniBlock, err := mbc.getGasUsedInMiniBlock(destMiniBlock)
```

Comment:
cache the results for the already calculated miniblocks. You are iterating over and over on the same set of data.